### PR TITLE
add simulateTransaction endpoint

### PR DESCRIPTION
**Pull request by tsachiherman**
## What's in this PR ?

This PR adds the simulateTransaction endpoint to the VM API.

## What are the changes in this PR ?

* The Recorder struct from the vmwithcontracts example was moved into the core of the hyper-sdk.
* Added simulateTransaction endpoint to the API server/client

## What is left and still need to be implemented in a subsequent PR ?

The existing support of `Simulate` in the `vmwithcontracts` remained as is. Removing it while taking advantage of the above would require some additional work that wasn't captured in this PR.

### CODE REVIEW

File: api/jsonrpc/server.go [hunk 1]
```diff
@@ -235,3 +236,96 @@ func (j *JSONRPCServer) Execute(
...
+}
+
+type SimulateTransactionReply struct {
+	SimulatedActions []SimulatedAction `json:"simulatedActions"`
+	ReadKeys         []string          `json:"readKeys"`
+	WriteKeys        []string          `json:"writeKeys"`
+	AllocateKeys     []string          `json:"allocKeys"`
+	BlockHeight      uint64            `json:"blockheight"`
+	Error            string            `json:"error"`
+}
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply api/jsonrpc/server.go:256]: I'd prefer to remove `BlockHeight` and `Error` (in favor of returning the error directly and pushing to the client to handle it) unless there's a clear need to include them in the response.

File: api/jsonrpc/server.go [hunk 2]
```diff
@@ -235,3 +236,96 @@ func (j *JSONRPCServer) Execute(
...
+	req *http.Request,
+	args *SimulateTransactionArgs,
+	reply *SimulateTransactionReply,
+) error {
+	ctx, span := j.vm.Tracer().Start(req.Context(), "JSONRPCServer.SimulateTransaction")
+	defer span.End()
+
+	if reply == nil {
+		return errors.New("SimulateTransaction was called with a nil reply object")
+	}
```
Comment (aaronbuchwald) [inline api/jsonrpc/server.go:276]: This should be provided by the server, so we can remove this check and follow the style in the rest of the APIs of not checking if the reply value is nil.

File: api/jsonrpc/client.go [hunk 3]
```diff
@@ -235,3 +236,55 @@ func Wait(ctx context.Context, interval time.Duration, check func(ctx context.Co
...
+	Error  error
+}
+
+type SimulateTransactionResponse struct {
+	SimulatedActions []SimulatedActionResponse
+	ReadKeys         []string
+	WriteKeys        []string
+	AllocateKeys     []string
+	BlockHeight      uint64
+}
```
Comment (aaronbuchwald) [inline api/jsonrpc/client.go:251]: Can we re-use the types specified in `server.go` as we do for the existing server/client implementations?

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]

File: chain/dependencies.go [hunk 4]
```diff
@@ -229,8 +238,8 @@ type Action interface {
 	// Execute actually runs the [Action]. Any state changes that the [Action] performs should
 	// be done here.
 	//
-	// If any keys are touched during [Execute] that are not specified in [StateKeys], the transaction
-	// will revert and the max fee will be charged.
+	// If any keys are touched during [Execute] while running in [OnchainTransactionExecution] mode that
+	// are not specified in [StateKeys], the transaction will revert and the max fee will be charged.
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply chain/dependencies.go]: done.

File: api/jsonrpc/server.go [hunk 5]
```diff
@@ -235,3 +236,96 @@ func (j *JSONRPCServer) Execute(
...
+
+func (SimulationEnabledRules) GetTransactionExecutionMode() chain.TransactionExecutionMode {
+	return chain.SimulatedTransactionExecution
+}
+
+func (j *JSONRPCServer) SimulateTransaction(
+	req *http.Request,
+	args *SimulateTransactionArgs,
+	reply *SimulateTransactionReply,
+) error {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go]: done

File: genesis/rules.go [hunk 6]
```diff
@@ -152,6 +152,10 @@ func (*Rules) FetchCustom(string) (any, bool) {
 	return nil, false
 }
 
+func (*Rules) GetTransactionExecutionMode() chain.TransactionExecutionMode {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply genesis/rules.go]: done.

File: api/jsonrpc/server.go [hunk 7]
```diff
@@ -235,3 +236,96 @@ func (j *JSONRPCServer) Execute(
...
 }
+
+type SimulateTransactionArgs struct {
+	Tx []byte `json:"tx"`
+}
+
+type SimulatedAction struct {
+	Output []byte  `json:"output_b64"`
+	Error  *string `json:"error"`
+}
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go]: done

File: api/jsonrpc/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go]: done

File: api/jsonrpc/server.go (hunk 2 above)
Comment (tsachiherman) [reply api/jsonrpc/server.go]: done

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]

File: examples/vmwithcontracts/vm/server.go [hunk 8]
```diff
@@ -101,7 +101,7 @@ func (j *JSONRPCServer) simulate(ctx context.Context, t actions.Call, actor code
 	if err != nil {
 		return nil, 0, err
 	}
-	recorder := storage.NewRecorder(currentState)
+	recorder := state.NewRecorder(currentState)
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply examples/vmwithcontracts/vm/server.go]: done

File: api/jsonrpc/client.go [hunk 9]
```diff
@@ -235,3 +237,36 @@ func Wait(ctx context.Context, interval time.Duration, check func(ctx context.Co
...
+
+	stateKeys := state.Keys{}
+	for _, entry := range resp.StateKeys {
+		hexBytes, err := hex.DecodeString(entry.HexKey)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		stateKeys.Add(string(hexBytes), state.Permissions(entry.Permissions))
+	}
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/client.go:270]: done

File: api/jsonrpc/server.go [hunk 10]
```diff
@@ -235,3 +237,71 @@ func (j *JSONRPCServer) Execute(
...
+) error {
+	ctx, span := j.vm.Tracer().Start(req.Context(), "JSONRPCServer.SimulateAction")
+	defer span.End()
+
+	if reply == nil {
+		return errors.New("SimulateAction was called with a nil reply object")
+	}
+
+	actionRegistry := j.vm.ActionRegistry()
+	rtx := codec.NewReader(args.Action, consts.NetworkSizeLimit) // will likely be much smaller than this
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go:269]: done

[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go:269]: done

File: state/recorder.go [hunk 11]
```diff
@@ -50,10 +48,10 @@ func (r *Recorder) GetValue(ctx context.Context, key []byte) (value []byte, err
 	return r.State.GetValue(ctx, key)
 }
 
-func (r *Recorder) GetStateKeys() state.Keys {
-	result := state.Keys{}
+func (r *Recorder) GetStateKeys() Keys {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/recorder.go:51]: the above isn't the correct example, but you're correct in saying that it won't perform correctly in the way it's currently implemented.

[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/recorder.go:51]: done

File: examples/vmwithcontracts/cmd/vmwithcontracts-cli/cmd/action.go [hunk 12]
```diff
@@ -141,9 +141,16 @@ var callCmd = &cobra.Command{
...
+			Fuel:            uint64(1000000000),
 		}
 
-		specifiedStateKeysSet, fuel, err := bcli.Simulate(ctx, *action, priv.Address)
+		specifiedStateKeysSet, encodedActionResults, err := cli.SimulateAction(ctx, action, priv.Address)
+		if err != nil {
+			return err
+		}
+
+		rtx := codec.NewReader(encodedActionResults, actions.MaxResultSizeLimit) // will likely be much smaller than this
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply examples/vmwithcontracts/cmd/vmwithcontracts-cli/cmd/action.go:152]: done

File: examples/vmwithcontracts/actions/call.go [hunk 13]
```diff
@@ -134,9 +139,32 @@ func (*Call) ValidRange(chain.Rules) (int64, int64) {
...
 
 func (*Result) GetTypeID() uint8 {
 	return mconsts.ResultOutputID
 }
+
+// Marshal encodes a type as bytes.
+func (r *Result) Marshal(p *codec.Packer) {
+	p.PackBytes(r.Value)
+	p.PackUint64(r.ConsumedFuel)
+}
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply examples/vmwithcontracts/actions/call.go:154]: done

File: api/jsonrpc/server.go [hunk 14]
```diff
@@ -235,3 +237,71 @@ func (j *JSONRPCServer) Execute(
...
+	actions, err := chain.UnmarshalActions(rtx, actionRegistry)
+	if err != nil {
+		return fmt.Errorf("%w: unable to unmarshal on public service", err)
+	}
+	if !rtx.Empty() {
+		return errors.New("tx has extra bytes")
+	}
+	if len(actions) != 1 {
+		return fmt.Errorf("simulateAction expects a single action, %d found", len(actions))
+	}
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go]: done

File: api/jsonrpc/server.go [hunk 15]
```diff
@@ -235,3 +237,71 @@ func (j *JSONRPCServer) Execute(
...
+			reply.Output, err = chain.MarshalTyped(actionOutput)
+			if err != nil {
+				return fmt.Errorf("failed to marshal output: %w", err)
+			}
+		}
+		if err != nil {
+			return err
+		}
+	}
+	for key, permission := range recorder.GetStateKeys() {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go]: done

File: api/jsonrpc/server.go [hunk 16]
```diff
@@ -235,3 +237,71 @@ func (j *JSONRPCServer) Execute(
...
+		reply.StateKeys = append(reply.StateKeys, SimulateStateKey{HexKey: hex.EncodeToString([]byte(key)), Permissions: byte(permission)})
+	}
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go]: sure. done

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]

File: api/jsonrpc/server.go [hunk 17]
```diff
@@ -233,44 +232,43 @@ func (j *JSONRPCServer) Execute(
...
+		}
+		if !actionsReader.Empty() {
+			return errors.New("tx has extra bytes")
+		}
+		actions = append(actions, action)
 	}
-	if len(actions) != 1 {
-		return fmt.Errorf("simulateAction expects a single action, %d found", len(actions))
+	if len(actions) == 0 {
+		return errors.New("simulateAction expects at least a single action, none found")
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go:271]: done

File: examples/vmwithcontracts/actions/call.go [hunk 18]
```diff
@@ -147,24 +147,8 @@ func (*Result) GetTypeID() uint8 {
...
-	p.PackBytes(r.Value)
-	p.PackUint64(r.ConsumedFuel)
-}
-
 // Size is the number of bytes it takes to represent this [Action]. This is used to preallocate
 // memory during encoding and to charge bandwidth fees.
 func (r *Result) Size() int {
 	return consts.Uint64Len + consts.Uint32Len + len(r.Value)
 }
-
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply examples/vmwithcontracts/actions/call.go]: done

File: api/jsonrpc/server.go [hunk 19]
```diff
@@ -230,3 +231,72 @@ func (j *JSONRPCServer) Execute(
...
+	actionRegistry := j.vm.ActionRegistry()
+	var actions chain.Actions
+	for _, actionBytes := range args.Actions {
+		actionsReader := codec.NewReader(actionBytes, len(actionBytes))
+		action, err := (*actionRegistry).Unmarshal(actionsReader)
+		if err != nil {
+			return err
+		}
+		if !actionsReader.Empty() {
+			return errors.New("tx has extra bytes")
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go:266]: done

File: api/jsonrpc/server.go [hunk 20]
```diff
@@ -230,3 +231,72 @@ func (j *JSONRPCServer) Execute(
...
+			if err != nil {
+				return fmt.Errorf("failed to marshal output: %w", err)
+			}
+		}
+		if err != nil {
+			return err
+		}
+		actionResult.StateKeys = recorder.GetStateKeys()
+		reply.ActionResults = append(reply.ActionResults, actionResult)
+		// create another recorder to recored the next action.
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go:298]: done

File: api/jsonrpc/server.go [hunk 21]
```diff
@@ -230,3 +231,72 @@ func (j *JSONRPCServer) Execute(
...
+		recorder = state.NewRecorder(recorder)
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply api/jsonrpc/server.go:299]: yes, done.

File: state/keys.go [hunk 22]
```diff
@@ -50,6 +56,40 @@ func (k Keys) ChunkSizes() ([]uint16, bool) {
...
+
+func (k Keys) MarshalJSON() ([]byte, error) {
+	var keysJSON keysJSON
+	for key, perm := range k {
+		keysJSON.Perms[perm] = append(keysJSON.Perms[perm], hex.EncodeToString([]byte(key)))
+	}
+	return json.Marshal(keysJSON)
+}
+
+func (k Keys) UnmarshalJSON(b []byte) error {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/keys.go:73]: I added the pointer, to align with the rest of the code, although it doesn't seems that it's needed.

[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/keys.go:73]: As for the JSON structure change, would it be ok if I'll create a separate issue for that ? it sounds sufficiently isolated and won't interfere with other aspects of this PR.

File: state/recorder.go [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+	"context"
+	"errors"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+// The Recorder struct wraps an [Immutable] state object and tracks the permissions used
+// against the various keys. The Recorder struct implements the [Mutable] interface, allowing
+// it to act as a direct replacement for a database view.
+// The Recorder struct maintain the same semantics as tstate_view in regards to the various
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/recorder.go:16]: done

File: state/recorder_test.go [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+		removedKeys = map[string]bool{}
+		// wrap with recorder.
+		recorder := state.NewRecorder(stateView)
+		for j := 0; j <= 32; j++ {
+			op := make([]byte, 1)
+			_, err := rand.Read(op)
+			require.NoError(err)
+			switch op[0] % 6 {
+			case 0: // insert into existing entry
+				randKey := randomKey()
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/recorder_test.go:81]: you're right. the function is mis-named. it should have been "pickExistingKeyAtRandom".

File: state/recorder_test.go [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+func (c *testingReadonlyDatasource) GetValue(_ context.Context, key []byte) (value []byte, err error) {
+	if v, has := c.storage[string(key)]; has {
+		return v, nil
+	}
+	return nil, database.ErrNotFound
+}
+
+func TestRecorderSideBySideFuzz(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/recorder_test.go:134]: it's definitely doable - but would take some time to implement. Would it be ok to use it as is for now and update these to use golang fuzzer in a subsequent PR ?

File: state/recorder.go [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+
+	changedValues map[string][]byte
+	keys          Keys
+}
+
+func NewRecorder(db Immutable) *Recorder {
+	return &Recorder{state: db, changedValues: map[string][]byte{}, stateKeys: map[string][]byte{}, keys: Keys{}}
+}
+
+func (r *Recorder) checkState(ctx context.Context, key []byte) error {
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply state/recorder.go]: done

**Review by aaronbuchwald — APPROVED** [review]

**Review by aaronbuchwald — APPROVED** [review]

**Pull request merged**